server/src/main/java/org/elasticsearch/action/bulk/Retry2.java

```diff
@@ -0,0 +1,297 @@
+
+    /*
+     * This is an approximation of the total number of bytes in the bulk requests that are with consumers.
+     */
+    private final AtomicLong totalBytesInFlight = new AtomicLong(0);
+    /*
+     * This is the approximate maximum number of bytes that this object will allow to be in flight to consumers.
+     */
+    private final long maxBytesInFlight;
+    /*
+     * We register in-flight calls with this Phaser so that we know whether there are any still in flight when we call awaitClose().
+     */
+    private final Phaser inFlightRequestsPhaser = new Phaser(1);
+
+    /**
+     * Creates a Retry2.
+     * @param maxNumberOfRetries This is the maximum number of times a BulkRequest will be retried
+     * @param maxBytesInFlight This is the maximum number of bytes that we want this object to keep in flight to the consumers
+     */
+    Retry2(int maxNumberOfRetries, ByteSizeValue maxBytesInFlight) {
+        this.logger = LogManager.getLogger(getClass());
+        this.maxNumberOfRetries = maxNumberOfRetries;
+        this.maxBytesInFlight = maxBytesInFlight.getBytes();
+    }
+
+    /**
+     * This method queues up the given BulkRequest. If there is no room on the queue, the listener is immediately notified of failure
+     * with an EsRejectedExecutionException. Otherwise, as soon as there is capacity consumer.accept(bulkRequest, actionListener) will be
+     * called (from another thread). If that call fails, the BulkRequest will be queued for retry based on this class's BackoffPolicy.
+     * @param consumer The consumer to which apply the request and listener. This consumer is expected to perform its work asynchronously
+     *                (that is, not block the thread from which it is called).
+     * @param bulkRequest The bulk request that should be executed.
+     * @param listener A listener that is invoked when the bulk request finishes or completes with an exception.
+     */
+    public void withBackoff(
+        BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer,
+        BulkRequest bulkRequest,
+        ActionListener<BulkResponse> listener
+    ) {
+        if (isClosing) {
+            listener.onFailure(new EsRejectedExecutionException("The bulk processor is closing"));
+            return;
+        }
+        List<BulkItemResponse> responsesAccumulator = new ArrayList<>();
+        logger.trace("Sending a bulk request with {} bytes in {} items", bulkRequest.estimatedSizeInBytes(), bulkRequest.requests.size());
+        /*
+         * Here we have a risk that multiple threads will do this check at once, and each will think it has room to run. So we could end
+         * up with slightly more bytes in flight than totalBytesInFlight. But this is acceptable risk, and preferable to using a lock or
+         * a while loop with compareAndSet.
+         */
+        long bytesInFlight = totalBytesInFlight.get();
+        if (bytesInFlight + bulkRequest.estimatedSizeInBytes() > maxBytesInFlight) {
```

Comment:
Could we do this check in `BulkProcessor2#add` rather than waiting until flush-time to reject things? Relatedly, could we count the size of the pending `BulkRequest` in `BulkProcessor2` towards the bytes-in-flight limit?

Reply:
Let me see what I can do. That means moving the tracking of bytes in flight to BulkProcessor2, which means a more complicated interaction between Retry2 and BulkProcessor2 (since we're subtracting bytes for the successful part of a request only when we have failures and then retry).

Reply:
I think it would be acceptable not to have such fine-grained tracking. We could just track all the bytes for the initial `BulkRequest` and hold on to them until it finally completes at which point they're all released.

Reply:
OK I've done this. Seems like a positive change, but it does make things a little more complicated.